signing_service/src/signing_service.py

```diff
@@ -0,0 +1,151 @@
+import argparse
+import logging.config
+import os
+import signal
+import socket
+from time import sleep
+
+from constants import SIGNING_SERVICE_DEFAULT_PORT
+from constants import SIGNING_SERVICE_MAXIMUM_RECONNECT_TIME
+
+
+logger = logging.getLogger()
+crash_logger = logging.getLogger('crash')
+
+
+class SigningService:
+    """
+    The Signing Service connects to Middleman as a client but then listens for requests coming from Concent via
+    Middleman. The underlying protocol is TCP and data sent over that is expected to conform to the Wire protocol.
+    """
+
+    def __init__(self, host, port, reconnect_delay) -> None:
+        assert isinstance(host, str)
+        assert isinstance(port, int)
+        assert isinstance(reconnect_delay, int)
+        self.host = host
+        self.port = port
+        self.reconnect_delay = reconnect_delay
+        self.current_reconnect_delay = 0
+        self.socket: socket.socket = None  # type: ignore
+        self.is_sigterm_sent = False
+
+    def run(self) -> None:
+        """
+        Handles main connection loop and its exceptions.
+
+        If connection is interrupted due to a failure, it should try to reconnect.
+        If connection is closed by other side, exit gracefully.
+        If a shutdown signal or KeyboardInterrupt is caught, exit gracefully.
+        If there was an unrecognized exception, it logs it and report to Sentry, then reraise and crash.
+        """
+
+        def _set_is_sigterm_sent_true(signum, frame):  # pylint: disable=unused-argument
+            self.is_sigterm_sent = True
+
+        # Handle shutdown signal.
+        signal.signal(signal.SIGTERM, _set_is_sigterm_sent_true)
+
+        while True:
+            if self.is_sigterm_sent:
+                self.socket.close()
+                logger.info('Closing connection and exiting on SIGTERM.')
+                break
+            try:
+                if self.socket is None:
+                    self._connect()
+                else:
+                    self.socket.send(b'')
```

Comment:
Why are you sending an empty string? I think there should be a comment here.

Reply:
Sending something is the only way to check if connection is still alive directly on the socket object. I will add a comment for this.

Reply:
:+1: done